numpyro/infer/mcmc.py

```diff
@@ -615,6 +621,257 @@ def __init__(self,
+        var = np.square(scale) + weight * np.square(new_sample - loc)
+        proposal_var = var - weight * np.square(samples - loc)
+        proposal_var = proposal_var - weight ** 2 * np.square(new_sample - samples)
+        proposal_scale = np.sqrt(proposal_var)
+
+    proposal_scale = np.concatenate([scale[None, ...], proposal_scale])
+    proposal_loc = loc + weight * (new_sample - samples)
+    proposal_loc = np.concatenate([loc[None, :], proposal_loc])
+    return proposal_loc, proposal_scale
+
+
+def _sample_proposal(inv_mass_matrix_sqrt, rng_key, batch_shape=()):
+    eps = random.normal(rng_key, batch_shape + np.shape(inv_mass_matrix_sqrt)[:1])
+    if inv_mass_matrix_sqrt.ndim == 1:
+        r = np.multiply(inv_mass_matrix_sqrt, eps)
+    elif inv_mass_matrix_sqrt.ndim == 2:
+        r = np.matmul(inv_mass_matrix_sqrt, eps[..., None])[..., 0]
+    else:
+        raise ValueError("Mass matrix has incorrect number of dims.")
+    return r
+
+
+SAAdaptState = namedtuple('SAAdaptState', ['zs', 'pes', 'loc', 'inv_mass_matrix_sqrt'])
+SAState = namedtuple('SAState', ['i', 'z', 'potential_energy', 'accept_prob',
+                                 'mean_accept_prob', 'diverging', 'adapt_state', 'rng_key'])
+
+
+# TODO: consider to expose this functional style
+def _sa(potential_fn=None, potential_fn_gen=None):
+    wa_steps = None
+    max_delta_energy = 1000.
+
+    def init_kernel(init_params,
+                    num_warmup,
+                    adapt_state_size=None,
+                    inverse_mass_matrix=None,
+                    dense_mass=False,
+                    model_args=(),
+                    model_kwargs=None,
+                    rng_key=PRNGKey(0)):
+        nonlocal wa_steps
+        wa_steps = num_warmup
+        pe_fn = potential_fn
+        if potential_fn_gen:
+            if pe_fn is not None:
+                raise ValueError('Only one of `potential_fn` or `potential_fn_gen` must be provided.')
+            else:
+                kwargs = {} if model_kwargs is None else model_kwargs
+                pe_fn = potential_fn_gen(*model_args, **kwargs)
+        rng_key_sa, rng_key_zs, rng_key_z = random.split(rng_key, 3)
+        z = init_params
+        z_flat, unravel_fn = ravel_pytree(z)
+        if inverse_mass_matrix is None:
+            inverse_mass_matrix = np.identity(z_flat.shape[-1]) if dense_mass else np.ones(z_flat.shape[-1])
+        inv_mass_matrix_sqrt = np.linalg.cholesky(inverse_mass_matrix) if dense_mass \
+            else np.sqrt(inverse_mass_matrix)
+        if adapt_state_size is None:
+            # XXX: heuristic choice
+            adapt_state_size = 2 * z_flat.shape[-1]
+        else:
+            assert adapt_state_size > 1, 'adapt_state_size should be greater than 1.'
+        # NB: mean is init_params
+        zs = z_flat + _sample_proposal(inv_mass_matrix_sqrt, rng_key_zs, (adapt_state_size,))
+        # compute potential energies
+        pes = lax.map(lambda z: pe_fn(unravel_fn(z)), zs)
```

Comment:
can you comment on which quantities are/are not cached from step to step?

Reply:
All information of the set S (the MCMC state of SA sampler) will be cached in `adapt_state`:
+ zs: values of the samples in S
+ pes: potential energies of them (we don't need to recompute those)
+ loc: mean of S
+ scale: scale of S